bh1750-control.1
.\"-
.\"Copyright (c) 2020, Alexander Mishin
.\"All rights reserved.
.\"
.\"Redistribution and use in source and binary forms, with or without
.\"modification, are permitted provided that the following conditions are met:
.\"
.\"* Redistributions of source code must retain the above copyright notice, this
.\"  list of conditions and the following disclaimer.
.\"
.\"* Redistributions in binary form must reproduce the above copyright notice,
.\"  this list of conditions and the following disclaimer in the documentation
.\"  and/or other materials provided with the distribution.
.\"
.\"THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
.\"AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\"IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
.\"DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
.\"FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\"DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
.\"SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
.\"CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
.\"OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
.\"OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
.Dd Sep 13, 2020
.Dt BH1750-CONTROL 1
.Os
.Sh NAME
.Nm bh1750-control
.Nd Ambient light control utility for bh1750 kernel driver
.Sh SYNOPSIS
.Nm Ar list
.Nm Ar add Fl l Ar level Fl e Ar command
.Op Fl s Ar scope
.Op Fl t Ar delay
.Nm Ar delete
.Op Fl s Ar scope
.Op Fl l Ar level
.Nm Ar run
.Sh OPTIONS
The following options are supported by
.Nm :
.Bl -tag -width automatic
.It Fl h
Prints a short help. Alse may be used for a help for a command.
.It Fl s Ar scope
Name for a whole set of defined level-command pairs. May be needed more
than one scope for one task. Default is "Default";
.It Fl l Ar level
Level of the light in millilux to add or delete a command definition;
.It Fl e Ar command
Command to execute when the illumination level in a certain area is
reached;
.It Fl t Ar delay
Command execution delay time in seconds. The command will be canceled
if during this time a change in illumination leads to a change of the
reached level in the defined scope. Default is 0.
.Sh EXAMPLES
The next lines specify a set of notifications for certain light levels
named "message" and a set for controlling gpio-pin 20 named "gpio" for
another list of light levels. In second case, for complete darkness, a
command execution delay of 60 seconds is set.

Then the ambient light level is measured, according to the result of
which two commands will be executed, one for each of the sets.

Finally, both sets are removed.
.Bd -literal
# bh1750-control add -s message -l 10000 -e "/bin/echo reached 10000nx"
# bh1750-control add -s message -l 20000 -e "/bin/echo reached 20000nx"
# bh1750-control add -s message -l 50000 -e "/bin/echo reached 50000nx"

# bh1750-control add -s gpio -l 0 -t 60 -e "/usr/sbin/gpioctl 20 0"
# bh1750-control add -s gpio -l 50000 -e "/usr/sbin/gpioctl 20 1"

# bh1750-control run

# bh1750-control delete -s message
# bh1750-control delete -s gpio
.Ed
.Sh SEE ALSO
.Xr bh1750 4 ,
.Xr bh1750-daemon 8 ,
.Sh AUTHORS
The
.Nm
was written by
.An Alexander Mishin Aq Mt mishin@mh.net.ru .


bh1750-control.py
#!/usr/local/bin/python3.7
# -*- coding: utf-8 -*-

import sys, subprocess
import sysctl, sqlite3, argparse

""" Class of illuminance sensor sysctl OID
"""
class illuminance():
    def __init__(self):
        self.oid = "dev.bh1750.0.illuminance"

    """ Get brightness level from sysctl """
    def get_level(self):
        try:
            level = sysctl.filter(self.oid)[0].value
        except Exception as e:
            if str(e) == "Invalid ctl_type: 0":
                exit("No bh1750.ko loaded.\n")
            else:
                raise e
        return level


""" Class for manipulate of actions DB
"""
class action_db():
    def __init__(self, filename):
        sql_create = (
            "CREATE TABLE IF NOT EXISTS scopes (\n"
            "    id INTEGER PRIMARY KEY AUTOINCREMENT,\n"
            "    scope TEXT NOT NULL UNIQUE\n"
            ");\n"
            "INSERT OR IGNORE INTO scopes VALUES(0,'Default');\n"
            "CREATE TABLE IF NOT EXISTS illuminance (\n"
            "    level INT NOT NULL,\n"
            "    scopeid INT NOT NULL,\n"
            "    delay INT NOT NULL DEFAULT 0,\n"
            "    action TEXT NOT NULL,\n"
            "    PRIMARY KEY (level, scopeid),\n"
            "    CONSTRAINT fk_scopes\n"
            "        FOREIGN KEY (scopeid)\n"
            "        REFERENCES scopes(id)\n"
            "        ON UPDATE CASCADE\n"
            "        ON DELETE CASCADE\n"
            ") WITHOUT ROWID;\n"
            "CREATE INDEX IF NOT EXISTS index_scope_level ON illuminance(scopeid, level);\n"
            "PRAGMA foreign_keys=ON;\n"
        )

        try:
            self.conn = sqlite3.connect(filename)
        except sqlite3.OperationalError as e:
            sys.stderr.write("%s\n" % e)

        self.cursor = self.conn.cursor()
        self.create_database(sql_create)

    """ Creates a database
    """
    def create_database(self, sql):
        self.cursor.executescript(sql)
        self.conn.commit()

    """ Adds an action for illuminance level and scope
    """
    def add(self, args):
        sql_add_scope = (
            "INSERT OR IGNORE INTO scopes(scope) VALUES(?);\n"
        )
        sql_add_level = (
            "INSERT INTO illuminance (level, scopeid, delay, action)\n"
            "VALUES (?, (SELECT id FROM scopes WHERE scope = ?), ?, ?);\n"
        )

        try:
            self.cursor.execute(sql_add_scope, (args.scope,))
            self.cursor.execute(sql_add_level, (args.level, args.scope, args.delay, args.execute,))
            self.conn.commit()
        except sqlite3.IntegrityError:
            sys.stderr.write(
                "Action on level:%slx for scope:\"%s\" already exists: Do You mean another scope?\n"
                % (args.level, args.scope)
            )
        except sqlite3.OperationalError as e:
            sys.stderr.write(
                "%s\n" % e
            )

    """ Deletes an action for scope and level if set
        or all actions for scope
    """
    def delete(self, args):
        level = args.level
        if level != None:
            sql_del = (
                "DELETE FROM illuminance\n"
                "WHERE level=? AND scopeid=(SELECT id FROM scopes WHERE scope=?);\n"
            )
            self.cursor.execute(sql_del, (level, args.scope,));
        else:
            sql_del = (
                "DELETE FROM scopes WHERE scope=?;\n"
            )
            self.cursor.execute(sql_del, (args.scope,));
        self.conn.commit()

    """ Get all actions for max illuminance level below
        specified one for all scopes
    """
    def run_actions(self, level):
        sql_get_actions = (
            "SELECT a1.level, a1.scopeid, a1.action FROM illuminance a1\n"
            "INNER JOIN (\n"
            "    SELECT MAX(level) AS max_level, scopeid\n"
            "    FROM illuminance WHERE level <= ? GROUP BY scopeid\n"
            ") a2\n"
            "ON a1.level = a2.max_level AND a1.scopeid = a2.scopeid;\n"
        )
        for level, scopeid, action in self.cursor.execute(sql_get_actions, (level,)):
            subprocess.run(action.split())

    """ Prints all action definitions from DB
    """
    def list_all(self):
        scope_prev = ""
        sql_list_all = (
            "SELECT level, scopeid, scope, delay, action FROM illuminance, scopes\n"
            "WHERE illuminance.scopeid = scopes.id\n"
            "ORDER BY scopeid, level\n"
        )

        for level, scopeid, scope, delay, action in self.cursor.execute(sql_list_all):
            if scope == scope_prev:
                sys.stdout.write("{:10d} {}\n".format(level, action))
            else:
                sys.stdout.write("[{}:{}]\n".format(scopeid, scope))
                if delay > 0:
                   sys.stdout.write("{:10d} {} (After: {} sec)\n".format(level, action, delay))
                else:
                   sys.stdout.write("{:10d} {}\n".format(level, action))
                scope_prev = scope


cmd_help = {
            "run"    : "execute the command for actual lighting level",
            "list"   : "list of all planned actions",
            "add"    : "add a command to execute for a lighting level in a scope",
            "delete" : "delete a planned action",
           }

parser = argparse.ArgumentParser(description='An illuminance level action')
sub_parsers = parser.add_subparsers(title="commands", dest="command")

for command in ["list", "run"]:
    p = sub_parsers.add_parser(command, 
                               help=cmd_help[command],
                               description=cmd_help[command].title())
for command in ["add", "delete"]:
    p = sub_parsers.add_parser(command,
                               help=cmd_help[command],
                               description=cmd_help[command].title())
    p.add_argument("-s", "--scope",
                   default="Default",
                   help="scope for row of levels and commands")
    if (command == "add"):
        p.add_argument("-t", "--delay",
                       default = 0,
                       help="seconds to delay before command")
        p.add_argument("-l", "--level",
                       required=True,
                       help="illuminance level for command to execute")
        p.add_argument("-e", "--execute",
                       required=True,
                       metavar="COMMAND",
                       help="command to execute")
    else:
        p.add_argument("-l", "--level",
                       help="illuminance level for command to execute")

db_filename = '/var/db/bh1750/actions.sqlite'

if __name__ == "__main__":
    try:
        action = action_db(db_filename)
    except sqlite3.OperationalError as e:
        sys.stderr.write("Operation not permitted: %s\n" % e)
        exit(-1)

    light = illuminance()

    args = parser.parse_args()

    if args.command == "list":
        action.list_all()
    elif args.command == "run":
        level = light.get_level()
        action.run_actions(level)
    elif args.command == "delete":
        action.delete(args)
    elif args.command == "add":
        action.add(args)

    action.conn.close()
